File: fallout.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>

int inarray(const char **array, int num, const char *string) {
	for (int i = 0; i < num; i++) {
		if (!strcmp(string, array[i])) {
			return 1;
		}
	}
	return 0;
}

void strtolower(char *ch) {
	for (; *ch; ch++) *ch = tolower(*ch);
}

int hamming(const char *op1, const char *op2) {
	int len = strlen(op1);
	int n = 0;
	for (int i = 0; i < len; i++) {
		if ( *(op1 + i) == *(op2 + i) ) n++;
	}

	return n;
}

int main(int argc, char **argv) {
	int i;
	int n_passw = argc - 1;

	if (n_passw < 2) {
		fprintf(stderr, "Too few arguments: %i\nUSAGE: %s PASSWORD PASSWORD...\n",
				n_passw, argv[0]);
		return 1;
	}

	char **passwords = &argv[1];
	int len = strlen(passwords[0]);
	for (i = 1; i < n_passw; i++) {
		if (strlen(passwords[i]) != len) {
			fprintf(stderr, "Passwords are of different length: '%s' and '%s'\n",
					passwords[0], passwords[i]);
			return 2;
		}
	}

	for (i = 0; i < n_passw; i++) strtolower(passwords[i]);

	char *input;
	int correct;
	for (i = 0; i < n_passw; i++) {
		printf("%s", "Enter password and number of correct letters\n> ");
		if ( scanf("%ms %i", &input, &correct) != 2 ) return 0;
		strtolower(input);
		if (!inarray((const char **) passwords, n_passw, input)) {
			fprintf(stderr, "%s is not in the list of possible passwords\n", input);
			return 3;
		}

		free(input);
	}
}

File: fallout.rb
#!/usr/bin/ruby
def hamming(op1, op2)
  i = 0
  op1.chars.zip(op2.chars) { |c| i += 1 if c[0] == c[1] }
  i
end

passwords = ARGV.map { |password| password.downcase }

if passwords.length < 1
  $stderr.puts <<~EOF
    Too few arguments: #{passwords.length}
    USAGE: #$0 PASSWORD PASSWORD...
  EOF
  exit 1
end

len_first = passwords[0].length
passwords.each do |password|
  if password.length != len_first
    $stderr.puts "Passwords are of different length: '%s' and '%s'" \
      % [passwords[0], password]
    exit 2
  end
end

1.upto passwords.length do
  print "Enter password and number of correct letters\n> "
  line = $stdin.gets
  if line.nil? then
    $stderr.puts "Empty input"
    exit 0
  end

  entered, correct = line.downcase.split /\W+/
  unless passwords.include? entered
    $stderr.puts "#{entered} is not in the list of possible passwords"
    exit 3
  end

  passwords.delete entered
  passwords.dup.each do |password|
    if hamming(password, entered) != correct.to_i then
      passwords.delete password
    end
  end

  num = passwords.length
  if num > 1 then
    puts "Possible passwords: #{passwords.join(' ')}"
  elsif num == 1 then
    puts "The password is: #{passwords[0]}"
    exit 0
  else
    $stderr.puts 'No more passwords left'
    exit 4
  end
end
